### PR TITLE
Improve the Prometheus parser error outputs to be more comprehensive

### PR DESCRIPTION
**Pull request by dgrisonnet**
Parsing errors in the Prometheus HTTP format parser are very hard to
investigate since they only approximately indicate what is going wrong
in the parser and don't provide any information about the incorrect
input. As such it is very hard to tell what is wrong in the format
exposed by the application.

I noticed the lack of information in the errors from the parser while investigating https://github.com/kubernetes-sigs/metrics-server/issues/1141. The problem we are facing is that we are reusing the internal parser instead of `expfmt` because the latter is significantly less performant than the one used internally by Prometheus and in some scenarios when metrics-server scrapes its target there is a parsing error which is hard to debug without having additional information in the errors.

This is most likely not a supported use-case but since this effort also improves the investigation of scraping errors in general in Prometheus, I thought it could be beneficial.

### CODE REVIEW

File: model/textparse/openmetricsparse.go
Comment (bwplotka) [inline model/textparse/openmetricsparse.go:228]: Can we have unified error formatting for OM, too or is it too much work for one PP?

**Review by bwplotka — APPROVED** [review]
Looks great to me, thanks!

Let's wait for last decision from maintainers of those particular packages cc @roidelapluie ?

File: model/textparse/openmetricsparse.go
Comment (dgrisonnet) [reply model/textparse/openmetricsparse.go:228]: I can add a TODO here and do it in a follow-up to make it easier to review

**Comment by roidelapluie** [comment]
cc @LeviHarrison what's your opinon here?

**Comment by LeviHarrison** [comment]
I really like this. I think I'd also add the line number in if that's possible, as it might help to know the full line for debugging. Also +1 to @bwplotka's comment about doing this for OM. Personally I don't think it would be an issue for reviewing if it was included in this PR too, but if the author's preference is to keep it separate I'm certainly fine with that.

**Comment by dgrisonnet** [comment]
> I really like this. I think I'd also add the line number in if that's possible, as it might help to know the full line for debugging.

I thought about that as well but considering that in most cases, the number of series exposed by the target will change between the time the it was scraped by Prometheus and the time a user tries to scrape it for debugging. This is why I made it so that the errors are expressive enough to give all the information needed for debugging.

**Comment by dgrisonnet** [comment]
Since everyone is fine with having the OM changes added to this PR, I'll add it when I get some time.

**Comment by LeviHarrison** [comment]
> I thought about that as well but considering that in most cases, the number of series exposed by the target will change between the time it was scraped by Prometheus and the time a user tries to scrape it for debugging. This is why I made it so that the errors are expressive enough to give all the information needed for debugging.

Fair point.

File: model/textparse/openmetricsparse.go
Comment (dgrisonnet) [reply model/textparse/openmetricsparse.go]: I added the error formatting to OM in my latest commit.

**Comment by dgrisonnet** [comment]
@LeviHarrison @bwplotka I added the error formatting to OM aswell, PTAL

**Review by LeviHarrison — APPROVED** [review]
LGTM. Just need to sort out tests.

**Comment by dgrisonnet** [comment]
Sorry about that, it should be fixed with the last update.

**Comment by LeviHarrison** [comment]
Thanks!

**Pull request merged**
